File: Interviews/GFG/Daily-Problems/2022/October/Day-06/Lexicographically-smallest-string.py
def lexicographicallySmallest(ob, S, k):
    n = len(S)
    if n & (n - 1):
        k <<= 1
    else:
        k >>= 1

    if k >= n: return "-1"
    ans = ""
    for c in S:
        while k > 0 and len(ans) and ans[-1] > c:
            ans = ans[:-1]
            k -= 1
        ans += c

    while k > 0:
        ans = ans[:-1]
        k -= 1

    return ans

File: Interviews/GFG/Daily-Problems/2022/October/Day-06/Lexicographically-smallest-string.cpp
string lexicographicallySmallest(string S, int k) {
    int n = S.size();
    if (n & (n - 1)) k <<= 1;
    else k >>= 1;

    if (k >= n) return "-1";
    string ans = "";
    for (int i = 0; i < n; i++) {
        while (k > 0 and ans.size() and ans.back() > S[i]) ans.pop_back(), k--;
        ans.push_back(S[i]);
    }

    while (k--) ans.pop_back();

    return ans;
}